Add particle handler

gassmoeller opened the pull request:
This PR illustrates how the classes in #4729 are planned to be used (and it is currently based on #4729). In particular it contains the structure and basic algorithms that are needed to generate, organize, communicate, and modify a collection of particles in a triangulation. The structure of `ParticleHandler` can be seen as analogous to `DoFHandler` with the difference that the `ParticleHandler` actually owns all of its data (i.e. there are no particle `solution` vectors stored externally). @luca-heltai: Some of the functionality we discussed is still missing (i.e. using deal.II's `find_active_cell_around_point` function and allowing to generate a set of particles from a given list of positions). I thought it simpler to add the basic structure first, and then extend the functionality.

This PR is currently lacking tests, I will add them during the next days.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+       * MPI communicator and the triangulation. This is useful after de-
+       * serialization of a particle handler.
+       */
+      void initialize(const parallel::distributed::Triangulation<dim,spacedim> &tria,
```
does it have to be `p::d::Tria`? what about `p::Tria`?

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+       * serialization of a particle handler.
+       */
+      void initialize(const parallel::distributed::Triangulation<dim,spacedim> &tria,
+                      const Mapping<dim,spacedim> &mapping,
```
maybe add some documentation in in the class or here in `initialize` to explain what `mappings` is for.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+       */
+      void initialize(const parallel::distributed::Triangulation<dim,spacedim> &tria,
+                      const Mapping<dim,spacedim> &mapping,
+                      const MPI_Comm mpi_communicator,
```
you probably don't need `MPI_Comm` as you can get it from `p::Tria` and if it's `Tria`, then set it to `MPI_COMM_SELF `

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+      /**
+       * This function allows to register three additional functions that are
+       * called every time a particle is transferred to another process
+       * (i.e. during sorting into cells, during ghost particle transfer, or
```
i think we need some further documentation in the class description to explain what `ghost particle` is. Most likely it's something owned by another MPI core, but this MPI core needs to know about it.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+       * respective cell center.
+       */
+      std::vector<std::vector<Tensor<1,spacedim> > >
+      vertex_to_cell_centers_directions(const std::vector<std::set<typename parallel::distributed::Triangulation<dim,spacedim>::active_cell_iterator> > &vertex_to_cells) const;
```
from the description, it does not look like this belongs to `ParticleHandler`. maybe it should be in `GridTools` or alike.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,480 @@
+       * neighbor of the current processor's domain owns this subdomain?
+       */
+      std::map<types::subdomain_id, unsigned int>
+      get_subdomain_id_to_neighbor_map() const;
```
same here, appears to be a general function outside of the `Particle` concept.

davydden commented:
I wonder if there is something one can do later to allow exchange of `ghost` particles when those are not associated to ghost cells. Say each MPI core needs to know about all particles within it's locally owned cells plus those which are within a certain distance to locally owned subdomain[1].

As an example, think of particles defining RHS in Laplace equation where each particle is a spherical function with local support. It would be great if whatever design decisions are done here would allow us later to cover this scenario.

[1] say we cover locally owned subdomain with `BoundingBox`es and therefore have a cheap way to tell if any particle outside of locally owned subdomain shall be a ghost for a given MPI core.

luca-heltai commented:
@gassmoeller , can you rebase on the current master, so that we have the particle class up to date, and only look at the `ParticleHandler` class?

gassmoeller commented:
I am currently travelling, so it will take another week or so (I also need to update with the latest updates in aspect, e.g. geodynamics/aspect#1948). I will come back to this as soon as possible.

gassmoeller commented:
I updated the pull request with the most recent status of the ParticleHandler class. Things I will do tomorrow:
- separate the fixes to the Particle class, and the changes to Triangulation into separate PRs
- create some test cases for the ParticleHandler class
- set up ASPECT to use the deal.II functionality if a recent enough deal.II is present. This will act as an additional test that future changes to deal.II do not change the interface (or if they do, we need to adjust ASPECT, but at least we know about it).

There are some things I would keep for follow-up PR. These include:
- the generalizations I discussed with @luca-heltai and @GivAlz (sending particles not only to neighbors, adding a flag for keeping particles in other domains, allowing to generate particles by a simple list of positions)
- working around some difficulties for the non-distributed triangulation case (the serialization will need to work differently for that)

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+     */
+    ParticleHandler(const parallel::distributed::Triangulation<dim,spacedim> &tria,
+                    const Mapping<dim,spacedim> &mapping,
+                    const MPI_Comm mpi_communicator,
```
why do you need extra MPI_Comm? You can get it from `p::d::Tria`.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+     * triangulation and the communicator are stored inside of the particle
+     *
+     */
+    ParticleHandler(const parallel::distributed::Triangulation<dim,spacedim> &tria,
```
you say you plant to extend this to `p::s::Tria`? Maybe take `p::Tria` here?

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+     *
+     */
+    ParticleHandler(const parallel::distributed::Triangulation<dim,spacedim> &tria,
+                    const Mapping<dim,spacedim> &mapping,
```
how is mapping used? Should probably be described above

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+    /**
+     * This function allows to register three additional functions that are
+     * called every time a particle is transferred to another process
+     * (i.e. during sorting into cells, during ghost particle transfer, or
```
what is ghost particle?

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+     * respective cell center.
+     */
+    std::vector<std::vector<Tensor<1,spacedim> > >
+    vertex_to_cell_centers_directions(const std::vector<std::set<typename parallel::distributed::Triangulation<dim,spacedim>::active_cell_iterator> > &vertex_to_cells) const;
```
from the description, it does not look like this belongs to ParticleHandler. maybe it should be in GridTools or alike.

davydden commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,497 @@
+     * neighbor of the current processor's domain owns this subdomain?
+     */
+    std::map<types::subdomain_id, unsigned int>
+    get_subdomain_id_to_neighbor_map() const;
```
same here, appears to be a general function outside of the Particle concept.

gassmoeller commented:
I addressed some of the comments and added a first test. I will continue next week. Most important task now is to add some more tests and make ASPECT run with this PR.

luca-heltai commented:
@gassmoeller , what's the status of this PR? @GivAlz and I are working on the next step we discussed, and we wanted to make sure we do the right thing... 

Can we talk on skype?

gassmoeller commented:
@luca-heltai I needed to fix some incompatibilities over the last two weeks, but that should be done by now. This PR is nearly ready, and at the top of my list. I sent you an email for Skype.

gassmoeller commented:
I think this would be a good state for this pull request to get another review (and maybe merge). I addressed most comments raised by @davydden (the parallel:shared:triangulation will have to wait a bit). I also added at least one test for every major functionality, and verified that the current version works like the one we currently use in aspect. 
@luca-heltai can you take a look at the `ParticleHandler::insert_particles(std::vector<Point<spacedim> >)` function we discussed? This is the most crude implementation imaginable, but it should do its job, and we can optimize it later on.
 When this is merged I will finish geodynamics/aspect#1957 to make sure the two versions of `ParticleHandler` do not get out of sync.

luca-heltai commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,495 @@
+    /**
+     * Insert a number of particles into the collection of particles.
+     * This function involves a copy of the particles and their properties.
+     * Note that this function is of O(n_existing_particles + n_particles) complexity.
```
Maybe add a note that this function expects all locations to be on locally owned cells (or do something in the .cc file when this is not the case, and document what is done in that case).

gassmoeller commented:
I addressed all of Luca's comments and added a test for the new function. Anything else to-do?

davydden commented:
i will have another look tomorrow morning...

jppelteret commented on include/deal.II/particles/particle.h:
```diff
@@ -25,63 +25,65 @@
 /**
- * A namespace that contains all classes that are related to the particle
- * implementation, in particular the fundamental Particle class.
+ * A namespace for all type definitions related to particles.
```
Can you please verify that this namespace description does not clash with [this one](http://dealii.org/developer/doxygen/deal.II/namespacetypes.html)? If so then I think that this description can simply be removed.

gassmoeller commented:
I moved the TODO into the implementation of the function, and removed the documentation of the namespace. I chose the namespace on purpose as the same one as for the other types, because a `particle_index` is conceptually the same thing as an `dof_index`. Would you prefer that I move the whole type declaration over to `types.h`? Everything else should be addressed.

gassmoeller commented:
Oh and could somebody start the tester so I can check that the tests are passing not just on my local machine?

Rombur commented:
/run-tests

davydden commented:
> I chose the namespace on purpose as the same one as for the other types, because a particle_index is conceptually the same thing as an dof_index. Would you prefer that I move the whole type declaration over to types.h? 

personally, i don't have a strong opinion on this. 

@dealii/dealii thoughts?

davydden commented:
build errors:
```
particle_handler.cc:300:5: error: no matching function for call to 'MPI_Scan'
```

jppelteret commented:
> I chose the namespace on purpose as the same one as for the other types, because a particle_index is conceptually the same thing as an dof_index. Would you prefer that I move the whole type declaration over to types.h?

I fine with extending the `dealii::types` namespace in the way that you have, especially since its use is more or less localised to files in the `particles` folder.

bangerth commented:
Yes, if we think of particles as a first class citizen in the deal.II universe, then it makes sense to put the appropriate data types in namespace `dealii::types`.

gassmoeller commented:
Hmm, so the MPI_Scan function on the tester expects different arguments than the one on my machines, and the one described on the OpenMPI website? `void *` instead of `const void*` as first argument? Anyway, I changed it, but now the tester again says `not allowed`. Is this because the test run is limited to a certain commit id? I amended my previous commit with the change.

From the rest of the answers I gathered that this is the last remaining issue.

bangerth commented:
The MPI 3.0 standard lists this as the correct signature:
```
int MPI_Scan(void* sendbuf, void* recvbuf, int count,
MPI_Datatype datatype, MPI_Op op, MPI_Comm comm)
```

gassmoeller commented:
Where did you find that? The report I found at http://mpi-forum.org/docs/mpi-3.0/mpi30-report.pdf states the following (and the same for MPI 3.1):
```
int MPI_Scan(const void* sendbuf, void* recvbuf, int count,
MPI_Datatype datatype, MPI_Op op, MPI_Comm comm)
```

Ah, I just checked MPI 2.2 (http://mpi-forum.org/docs/mpi-2.2/mpi22-report.pdf), there it is:
```
int MPI_Scan(void* sendbuf, void* recvbuf, int count,
MPI_Datatype datatype, MPI_Op op, MPI_Comm comm )
```

Anyway, the new version should work in both cases.

bangerth commented:
Ah, how bizarre -- it turns out that I only had a draft among my documents, and that draft did not have the `const` in it. Apparently I need to update the draft :-)

bangerth commented:
All the failing tests are due to timeouts.

bangerth commented on include/deal.II/particles/particle_handler.h:
```diff
@@ -0,0 +1,507 @@
+
+#ifdef DEAL_II_WITH_MPI
+    /**
+     * Exchanges all particles that live in cells that are ghost cells to
```
I would prefer if the declaration was always available, and only the *body* of the implementation is commented out when not using MPI. This way, the function just does nothing, but an MPI code can also be compiled without using MPI.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -0,0 +1,1152 @@
+
+#ifdef DEAL_II_WITH_MPI
+  template <int dim, int spacedim>
+  void
```
This is the place where you'd move the `#ifdef` down into the body of the function.

bangerth commented on source/particles/particle_handler.cc:
```diff
@@ -0,0 +1,1152 @@
+
+    send_recv_particles(ghost_particles_by_domain,
+                        ghost_particles);
+  }
```
and you need the `#endif` here, and another `#ifdef` in front of the next function, since that is definitely commented out. (Without harm in the latter case since it's a private function of the class. So it can't be called from anywhere else, and it's not called from within this class if we're not using MPI.)

bangerth commented:
I've looked through this again and it looks like it's definitely ready. I believe that's also what the others mention above. So if you could address the minor issues still pointed out a few minutes ago, I'll merge.

gassmoeller commented:
I addressed the issues.

bangerth commented:
Thanks, @gassmoeller! The four failing tests are timeouts again, so I'll merge.
